File: 2461-Maximum Sum of Distinct Subarrays With Length K.py
class Solution(object):
    def maximumSubarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        using = collections.defaultdict(int)
        res = 0
        left = 0
        curr = 0
        
        for j in range(len(nums)):
            if j >= k:
                curr -= nums[j-k]
                using[nums[j-k]] -= 1
            if using[nums[j-k]] == 0:
                del using[nums[j-k]]
            
            curr += nums[j]
            using[nums[j]] += 1
            if len(using) == k:
                res = max(res, curr)        
        return res

File: C++/2461-Maximum Sum of Distinct Subarrays With Length K.cpp
class Solution {
public:
    long long maximumSubarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> cnt;
        long long res = 0, tot = 0;
        for (int i=0; i<k; i++) {
            cnt[nums[i]]++;
            tot += nums[i];
        }

        if (cnt.size() == k) res = tot;

        for (int i=k; i<nums.size(); i++) {
            tot += nums[i]-nums[i-k];
            cnt[nums[i]]++;
            if (--cnt[nums[i-k]] == 0) cnt.erase(nums[i-k]);
            if (cnt.size() == k) {
                res = max(res, tot);
            }
        }
        return res;
    }
};